### ISSUE TITLE
F2PY: Allow the use of "iso_fortran_env" in pyf signature declarations

### ISSUE THREAD

**Issue by MarkWieczorek** [numpy/numpy#17563]
## Declaration of fortran types in pyf signtature file using "use iso_fortran_env"

When calling fortran functions in python it is necessary to create both a wrapped function as well as a python signature file (pyf). In the fortran wrapper function, most people make use of the `iso_fortran_env` module to define their types, like this:

```
 # wrapper function
   subroutine pyfunc(i, x)
        use iso_fortran_env, only: sp=>real32, dp=>real64, qp=>real128
        use iso_fortran_env, only: int8, int16, int32, int64
        implicit none
        integer(int32),intent(in) :: i
        real(dp),intent(out) :: x
        call func(i, x)
    end subroutine pyfunc
```

However, in the python signature file, it is not possible to use `iso_fortran_env`. I have tried to use this module, but f2py simply ignores it. According to the documentation [online](https://numpy.org/devdocs/f2py/signature-file.html): "Currently F2PY uses use statement only for linking call-back modules and external arguments (call-back functions)", and this suggests to me that using arbitrary modules is not implemented.

Since it is not possible to use `iso_fortran_env` in the signature file, one possibility is to instead use intrinsic functions like `selected_real_kind`

```
# python signature file
python module sig
    interface

        subroutine func(i, x)
            fortranname pyfunc
            integer, parameter :: dp = selected_real_kind(p=15)
            integer, parameter :: int32 = selected_int_kind(9)
            integer(int32),intent(in) :: i
            real(dp),intent(out) :: x
        end subroutine func

    end interface
end python module sig
```
This works, but it is inconsistent, and could conceivably fail with some compilers/architectures. In particular, in the wrapper function we are defining the kind by the number of bits, but in the signature file by the precision of the number.

I note that there is a very short discussion [here](https://numpy.org/devdocs/f2py/advanced.html#dealing-with-kind-specifiers) on how to modify the kind declarations to be compatible with C-types. However, if there is an easy way to do this, it is not obvious from the extremely terse and undocumented nature of these features.

In conclusion: For simplicity, the types defined in the `iso_fortran_env` module should be made accessible to the python signature file.

**Comment by 2sn** [comment]
May I add to this the following observation:
the
```
  use, intrinsic :: &
       iso_fortran_env
```
seems to fail only in subroutines contained in a module.  Potentially I am doing something wrong.  Please consider the following code (that compiles for me with numpy 1.21.4):
```

    ! use, intrinsic :: &
    !      iso_fortran_env

    implicit none

    save

    INTEGER, PARAMETER :: int32 = SELECTED_INT_KIND(8)
    INTEGER, PARAMETER :: real64 = SELECTED_REAL_KIND(15)

    real(kind=real64), intent(in), dimension(:,:) :: &
         v1_, v2_
    real(kind=real64), intent(in), dimension(:) :: &
         t1_, t2_
    real(kind=real64), intent(in) :: &
         toffset_

    if (allocated(t1)) then
       deallocate(v1, v2, t1, t2)
    endif

    n1 = size(t1_, 1)
    allocate(t1(n1))
    if (n1 /= size(v1_, 1)) STOP '[SET_STARDATA] dimension error 1'
    m1 = size(v1_, 2)
    allocate(v1(n1, m1))

    n2 = size(t2_, 1)
    allocate(t2(n2))
    if (n2 /= size(v2_, 1)) STOP '[SET_STARDATA] dimension error 2'
    m2 = size(v2_, 2)
    allocate(v2(n2, m2))

    toffset=toffset_
    t1 = t1_
    t2 = t2_
    v1 = v1_
    v2 = v2_

  end subroutine set_stardata

end module stardata

subroutine set_stardata1(t1_, v1_, t2_, v2_, toffset_)

  ! use, intrinsic :: &
  !      iso_fortran_env

  use stardata

  implicit none

  save

  ! INTEGER, PARAMETER :: int32 = SELECTED_INT_KIND(8)
  ! INTEGER, PARAMETER :: real64 = SELECTED_REAL_KIND(15)

  real(kind=real64), intent(in), dimension(:,:) :: &
       v1_, v2_
  real(kind=real64), intent(in), dimension(:) :: &
       t1_, t2_
  real(kind=real64), intent(in) :: &
       toffset_

  if (allocated(t1)) then
     deallocate(v1, v2, t1, t2)
  endif

  n1 = size(t1_, 1)
  allocate(t1(n1))
  if (n1 /= size(v1_, 1)) STOP '[SET_STARDATA] dimension error 1'
  m1 = size(v1_, 2)
  allocate(v1(n1, m1))

  n2 = size(t2_, 1)
  allocate(t2(n2))
  if (n2 /= size(v2_, 1)) STOP '[SET_STARDATA] dimension error 2'
  m2 = size(v2_, 2)
  allocate(v2(n2, m2))

  toffset=toffset_
  t1 = t1_
  t2 = t2_
  v1 = v1_
  v2 = v2_

end subroutine set_stardata1
```
In the external routine `set_stardata1` I run into trouble when uncommenting the explicit type definition but no ham uncommenting the use statement (both is as it should be).

But in the inner routine, the direct definition seems to be required, commenting it out gives errors
```
/tmp/tmp0kwpa4or/src.linux-x86_64-3.9/_solver-f2pywrappers2.f90:159:16:

  159 |       real(kind=real64) toffset_
      |                1
Error: Parameter ‘real64’ at (1) has not been declared or is a variable, which does not reduce to a constant expression
/tmp/tmp0kwpa4or/src.linux-x86_64-3.9/_solver-f2pywrappers2.f90:166:16:

  166 |       real(kind=real64) t1_(f2py_t1__d0)
      |                1
Error: Parameter ‘real64’ at (1) has not been declared or is a variable, which does not reduce to a constant expression
/tmp/tmp0kwpa4or/src.linux-x86_64-3.9/_solver-f2pywrappers2.f90:167:16:

  167 |       real(kind=real64) v1_(f2py_v1__d0,f2py_v1__d1)
      |                1
Error: Parameter ‘real64’ at (1) has not been declared or is a variable, which does not reduce to a constant expression
/tmp/tmp0kwpa4or/src.linux-x86_64-3.9/_solver-f2pywrappers2.f90:168:16:

  168 |       real(kind=real64) t2_(f2py_t2__d0)
      |                1
Error: Parameter ‘real64’ at (1) has not been declared or is a variable, which does not reduce to a constant expression
/tmp/tmp0kwpa4or/src.linux-x86_64-3.9/_solver-f2pywrappers2.f90:169:16:

  169 |       real(kind=real64) v2_(f2py_v2__d0,f2py_v2__d1)
      |                1
Error: Parameter ‘real64’ at (1) has not been declared or is a variable, which does not reduce to a constant expression
/tmp/tmp0kwpa4or/src.linux-x86_64-3.9/_solver-f2pywrappers2.f90:187:16:

  187 |       real(kind=real64) toffset_
      |                1
Error: Parameter ‘real64’ at (1) has not been declared or is a variable, which does not reduce to a constant expression
/tmp/tmp0kwpa4or/src.linux-x86_64-3.9/_solver-f2pywrappers2.f90:194:16:

  194 |       real(kind=real64) t1_(f2py_t1__d0)
      |                1
Error: Parameter ‘real64’ at (1) has not been declared or is a variable, which does not reduce to a constant expression
/tmp/tmp0kwpa4or/src.linux-x86_64-3.9/_solver-f2pywrappers2.f90:195:16:

  195 |       real(kind=real64) v1_(f2py_v1__d0,f2py_v1__d1)
      |                1
Error: Parameter ‘real64’ at (1) has not been declared or is a variable, which does not reduce to a constant expression
/tmp/tmp0kwpa4or/src.linux-x86_64-3.9/_solver-f2pywrappers2.f90:196:16:

  196 |       real(kind=real64) t2_(f2py_t2__d0)
      |                1
Error: Parameter ‘real64’ at (1) has not been declared or is a variable, which does not reduce to a constant expression
/tmp/tmp0kwpa4or/src.linux-x86_64-3.9/_solver-f2pywrappers2.f90:197:16:

  197 |       real(kind=real64) v2_(f2py_v2__d0,f2py_v2__d1)
      |                1
Error: Parameter ‘real64’ at (1) has not been declared or is a variable, which does not reduce to a constant expression
/tmp/tmp0kwpa4or/src.linux-x86_64-3.9/_solver-f2pywrappers2.f90:170:53:

  170 |       call set_stardata(t1_, v1_, t2_, v2_, toffset_)
      |                                                     1
Error: Type mismatch in argument ‘t1_’ at (1); passed REAL(4) to REAL(8)
```
Same just commenting out the use statement - the use intrinsic statement inside the module procedure does not seem to work; it works on the module level itself as well as outside the module.  Maybe there is an issue with the level of nesting in `f2py`?

Even explicit definition on the module level is ignored insider the module subroutine, and a redefinition inside the routine  is required.

**Comment by 2sn** [comment]
An adequately usable solution may be to define my own tupe module  ... 
```
 module types

  implicit none

  save

  INTEGER, PARAMETER :: int32 = SELECTED_INT_KIND(8)
  INTEGER, PARAMETER :: real64 = SELECTED_REAL_KIND(15)

end module types



module stardata

  use types

  implicit none

  save

  real(kind=real64), allocatable, dimension(:,:) :: &
       v1, v2
  real(kind=real64), allocatable, dimension(:) :: &
       t1, t2

  real(kind=real64) :: &
       toffset = 0.d0

  integer(kind=int32) :: &
       n1 = -1, &
       n2 = -1, &
       m1 = -1, &
       m2 = -1


contains

  subroutine set_stardata(t1_, v1_, t2_, v2_, toffset_)

    use types

    implicit none

    save

    real(kind=real64), intent(in), dimension(:,:) :: &
         v1_, v2_
    real(kind=real64), intent(in), dimension(:) :: &
         t1_, t2_
    real(kind=real64), intent(in) :: &
         toffset_

    if (allocated(t1)) then
       deallocate(v1, v2, t1, t2)
    endif

    n1 = size(t1_, 1)
    allocate(t1(n1))
    if (n1 /= size(v1_, 1)) STOP '[SET_STARDATA] dimension error 1'
    m1 = size(v1_, 2)
    allocate(v1(n1, m1))

    n2 = size(t2_, 1)
    allocate(t2(n2))
    if (n2 /= size(v2_, 1)) STOP '[SET_STARDATA] dimension error 2'
    m2 = size(v2_, 2)
    allocate(v2(n2, m2))

    toffset=toffset_
    t1 = t1_
    t2 = t2_
    v1 = v1_
    v2 = v2_

  end subroutine set_stardata

end module stardata
```
This seems to work the way `iso_fortran_env` was supposed to work.
